packages/firestore/src/remote/bloom_filter.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Md5, Integer } from '@firebase/webchannel-wrapper';
+
+import { debugAssert } from '../util/assert';
+
+export class BloomFilter {
+  private readonly bitSize: number;
+
+  constructor(
+    private readonly bitmap: Uint8Array,
+    padding: number,
+    private readonly hashCount: number
+  ) {
+    this.bitSize = this.bitmap.length * 8 - padding;
+    if (this.bitSize === 0) {
+      debugAssert(
+        this.hashCount === 0 && padding === 0,
+        'A valid empty bloom filter should have all 3 fields empty.'
+      );
+    } else {
+      debugAssert(padding >= 0, 'Padding is negative.');
+      debugAssert(this.bitSize > 0, 'Bitmap size is negative.');
+      debugAssert(this.hashCount > 0, 'Hash count is 0 or negative');
+    }
+  }
+
+  getBitSize(): number {
+    return this.bitSize;
+  }
+
+  mightContain(document: string): boolean {
+    // Empty bitmap should always return false on membership check
+    if (this.bitSize === 0) {
+      return false;
+    }
+
+    // Hash the string using md5
+    const md5 = new Md5();
+    md5.update(document);
+    const encodedBytes: Uint8Array = new Uint8Array(md5.digest());
+
+    // Interpret the hashed value as two 64-bit chunks as unsigned integers, encoded using 2’s
+    // complement using little endian.
+    const dataView = new DataView(encodedBytes.buffer);
+    const firstUint32 = dataView.getUint32(0, /* littleEndian= */ true);
+    const secondUint32 = dataView.getUint32(4, /* littleEndian= */ true);
+    const thirdUint32 = dataView.getUint32(8, /* littleEndian= */ true);
+    const fourthUint32 = dataView.getUint32(12, /* littleEndian= */ true);
+    const hash1 = new Integer([firstUint32, secondUint32], 0);
+    const hash2 = new Integer([thirdUint32, fourthUint32], 0);
+
+    for (let i = 0; i < this.hashCount; i++) {
+      // Calculate hashed value h(i) = h1 + (i * h2), wrap if hash value overflow
+      let combinedHash = hash1.add(hash2.multiply(Integer.fromNumber(i)));
+      combinedHash = combinedHash.modulo(Integer.fromNumber(Math.pow(2, 64)));
```

Comment:
Save `Integer.fromNumber(Math.pow(2, 64))` into a module-level constant so that it doesn't need to be recalculated all the time. Also, "modulo" is a pretty expensive operation. I think that what you *really* want is to just drop the leftmost bits, leaving just 64 bits. I think you could use the "getBitsUnsigned( index )" method of Integer to just grab the 64 rightmost bits and create a new Integer from it. That would be computationally much cheaper. Can you look into that?